torch/_inductor/compile_fx_async.py

```diff
@@ -179,3 +179,183 @@ def callback(pickled_output: _WireProtocolPickledOutput) -> OutputCode:
+    _post_compile_data: Optional[_PostCompileData] = None
+    _boxed_call: bool
+    _current_progression_index: int
+
+    def __init__(
+        self,
+        # Fast compile that runs faster than the progressive compiles
+        fast_output_code: OutputCode,
+        # Futures for the progressive optimized compiles
+        progression_futures: list[Future[_WireProtocolPickledOutput]],
+        # Callback to convert the optimized result to OutputCode
+        callback: Callable[[_WireProtocolPickledOutput], OutputCode],
+    ) -> None:
+        self._fast_output_code = fast_output_code
+        self._optimized_output_code = None
+        self._progression_futures = progression_futures
+        self._callback = callback
+        self._boxed_call = getattr(fast_output_code, "_boxed_call", False)
+        self._current_progression_index = -1
+
+    @override
+    def __call__(self, *args: Any) -> Any:
+        # Check if any newer progression stage is ready and switch to it
+        args = self._check_and_switch_progression(args)
+
+        if self._optimized_output_code is not None:
+            _ProgressiveFxCompile._stat_optimized_runs += 1
+            return self._optimized_output_code.__call__(*args)
+        else:
+            _ProgressiveFxCompile._stat_fast_runs += 1
+            assert self._fast_output_code is not None
+            return self._fast_output_code.__call__(*args)
+
+    def _check_and_switch_progression(self, args: tuple[Any, ...]) -> tuple[Any, ...]:
+        # Check if any newer progression stage is ready (in order from latest to earliest)
+        for i in range(
+            len(self._progression_futures) - 1, self._current_progression_index, -1
+        ):
+            future = self._progression_futures[i]
+            if future.done():
+                args = self._switch_to_progression_stage(i, args)
+                break
+
+        return args
+
+    def _switch_to_progression_stage(
+        self, stage_index: int, args: tuple[Any, ...]
+    ) -> tuple[Any, ...]:
+        future = self._progression_futures[stage_index]
+        optimized_output_code = self._callback(future.result())
+
+        if pcd := self._post_compile_data:
+            self._post_compile_data = None
+            optimized_output_code.post_compile(
+                pcd.example_inputs, pcd.constants, pcd.graph_kwargs
+            )
+
+        self._optimized_output_code = optimized_output_code
+        self._fast_output_code = None
+        self._current_progression_index = stage_index
+
+        optimized_boxed_call = getattr(optimized_output_code, "_boxed_call", False)
+
+        if self._boxed_call != optimized_boxed_call:
+            if self._boxed_call:
+                # Was boxed, now unboxed
+                args = args[0] if len(args) > 0 else ()
+            else:
+                # Was unboxed, now boxed
+                args = (args,)
+
+        self._boxed_call = optimized_boxed_call
```

Comment:
I thought in one of my PRs I changed this because something else was caching the fact that our function was boxed or not so we couldn't dynamically change it so I just stated that we were always boxed and we just converted if the callee wasn't.